cmd/thanos/receive.go

```diff
@@ -97,13 +134,45 @@ func runReceive(
 		MaxBlockDuration:  model.Duration(time.Hour * 2),
 	}
 
+	var hashring receive.Hashring
+	// Run File Service Discovery and check for updates.
+	if fileSD != nil {
+		ctxRun, cancelRun := context.WithCancel(context.Background())
+		fileSDUpdates := make(chan []*targetgroup.Group)
+		go fileSD.Run(ctxRun, fileSDUpdates)
+
+		// We need to wait for initial discovery to create the hashring.
+		groups := <-fileSDUpdates
+		hashring = receive.NewHashring(receive.ExactMatcher, groups)
+		g.Add(func() error {
+			for {
+				select {
+				// If the SD changes, we need to shutdown.
+				case update := <-fileSDUpdates:
+					if !receive.HashringsAreEquivalent(groups, update) {
+						level.Info(logger).Log("msg", "file service discovery changed")
+						return nil
```

Comment:
Is shutting down really the right thing to do here? I feel like not accepting writes would be more appropriate, as kubernetes or equivalent would just restart the process, and really what we expect is that something from the outside initiates a shutdown causing flushing + upload with a subsequent rollout.

Reply:
updated, now the handler will simply stop accepting write-requests when the hashring changes.